Flexible schema config in yaml

bboreham opened the pull request:
Fixes #1014 
Fixes #320 

The schema config in yaml looks like this:

```
configs:
- from: 0
  store: aws
  schema: v1
  index:
    prefix: dev_chunks_2
- from: 1484006400000
  store: aws
  schema: v2
  index:
    prefix: dev_chunks_2
- from: 1486339200000
  store: aws
  schema: v4
  index:
    prefix: dev_chunks_weekly_
    period: 168h
- from: 1502928000000
  store: aws-dynamo
  schema: v9
  index:
    prefix: dev_chunk_index_
    period: 168h
  chunks:
    prefix: dev_chunk_data_weekly_
    period: 168h
- from: 1508371200000
  store: aws-dynamo
  schema: v6
  index:
    prefix: dev_chunks_weekly_
    period: 168h
  chunks:
    prefix: dev_chunk_data_weekly_
    period: 168h
```

If you don't specify a `Period` then all data goes in one table and `prefix` is the entire name of that table.

At present `model.Time` does not unmarshal from a human-readable format.

`StorageOpts` are gone: all the work to create objects from the config is done in `storage/factory`. ~I would prefer that `CompositeStoreEntry` did not get exported, but the collusion between chunk and storage requires it.~

`schema` and `compositeStoreEntry` are still somewhat parallel - I might fix that later.

The legacy command-line option for the storage client - "aws", "cassandra", etc., moves from storage package into the schema area, because it is used during initialization.

I haven't squashed commits at all.  Includes the change in #1063

bboreham commented:
Oh, and `table-manager` isn't quite finished.  I'm focused on the query side first.

bboreham commented:
Another improvement would be to watch for changes in the config and reload without having to restart the process.

leth commented:
Personally I'd find it easier to parse a textual datestamp in the `from` field - I'd be far less likely to make a mistake reviewing a date change, rather than a timestamp change.

bboreham commented:
> Personally I'd find it easier to parse a textual datestamp in the `from` field

That's what this line relates to:

> At present `model.Time` does not unmarshal from a human-readable format.

I will probably have to use a different type in the config.

bboreham commented:
Rebased off current master and squashed.

I left in a `PrintYaml()` routine and associated `clean()` functions which aren't called - I hacked `querier/main.go` to create my first yaml file.  Any better ideas for how to do that?
